Pulsed source

dougUCN opened the pull request:
This pull request does the following:

1. Prevents CMAKE_BUILD_TYPE from being 'Debug' to avoid CGAL compilation warning (One line change to CMakeLists.txt)
2. Fixes an error in the unit tests: "unknown location(0): fatal error: in TFieldManagerTest: std::runtime_error: Section FORMULAS not found in config file" (One line change to test/fieldTests.cpp)
3. Adds

wschreyer commented:
Thanks.

To 1.) You should not override the Debug build flag in the CMakeLists in case someone actually wants to create a debug build. You can define CMAKE_BUILD_TYPE as a command line parameter for cmake to create any build you want. I included the CMAKE_BUILD_TYPE line only to make sure that it will be a "Release" build if the user does not define any cmake parameters.

To 3.) I had been thinking about including time-dependent sources but never had an application for it. Although I was considering using a more general user-defined formula instead.
You could simplify your solution by creating the piecewise_constant_distribution in the TParticleSource constructor and then just calling the distribution generator in all the CreateParticle functions (similar to the spectrum and angular distributions). You should be able to use piecewise_constant_distribution for all eventualities: single pulse of length ActiveTime, multiple pulses, or even pulses with length 0 (see here how to handle that case: https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution/uniform_real_distribution, I realize now this case could actually lead to undefined behavior in the original code)

dougUCN commented:
1) Ahhh alright, I will undo that change with a commit

3) I considered trying to create a user-defined time dependent source, but had difficulty trying to figure out a way to convert an expression like "t%5 = 0" into an arbitrary probability distribution.  I also wanted to avoid the user having to write something like "0-5 10-15 20-25..." since my desired use case is to simulate the pulsed Lanl ucn source.

In an earlier commit, 2298c5d34eac370ad5c4263e91f39e9ae845ea5f 
I have indeed utilized piecewise_constant_distribution for all situations as you suggested, and defined it in the constructor.

However, for some inexplicable reason, this would cause PENTrack to freeze during particle generation very often, and the simulation ran noticeably slower. I don't understand why it would do so.  After trying to troubleshoot this for a few days and writing several different test scripts on the behavior of piecewise_constant_distribution, I gave up and implemented the method in the most recent commit. I agree that it certainly isn't a very clean solution.

wschreyer commented:
Yeah, a non-continuous time distribution would not be very well reflected using a user-defined formula. For spectra etc. I simply interpolate the formula with a piecewise_linear_distribution over a lot of points.
I wonder if the edge case is to blame for the issues you had? Like a situation where your last pulse starts when ActiveTime ends might lead to a situation where your last pulse has a length 0.

dougUCN commented:
I saw your spectra distribution solution, that was very elegant, yes

The PENTrack freeze issue occurred even in cases with just a uniform distribution across a fixed interval or if I started all UCN at t=0. This was tested on the base Penelope example as well as my own simulations that previously ran without issue.

In regards to that last edge case, currently my code sets the probability weighting of that interval to 0. I will further stress test it but I don't think that was the root cause to my issue.

wschreyer commented:
I did a quick checkout of 2298c5d34eac370ad5c4263e91f39e9ae845ea5f on my WSL Ubuntu and ran a couple of different settings (including ActiveTime 0 and pulseWidth 0) in the default config and ran into no issues.
You could try a clean recompile (rm -rf CMakeCache CMakeFiles && cmake .) and see if the issue persists.

Unfortunately the doc for piecewise_constant_distribution does not say what happens in those edge cases, but if it is undefined behavior as with the uniform_distribution it might be compiler dependent.

dougUCN commented:
Oh really? That's good to hear. 

I did only `make clean`, will try clearing the cmake cache and report back to you on this issue tomorrow.

dougUCN commented:
I have implemented the solution discussed above to handle edge cases, etc. See 8c1a86f81eb1e16d0e05aa110449dd0c295d997f. Also did a clean recompile and I think everything is running okay. Unit tests passed as well. Let me know if everything runs okay on your end as well.

Additionally, I found that the comsol field parameter descriptions in PENTrack/in/config.in are incorrect, as there is no Efieldscale implemented in field_3d for comsol fields. Corrected this documentation in commit 303450e15e262ec00dc04a7afc86339d6db810ab

wschreyer commented:
Looks good, thanks! Sorry it took me so long.
